modelforge/potential/sake.py

```diff
@@ -0,0 +1,382 @@
+        )
+
+        self.post_norm_mlp = nn.Sequential(
+            Dense(self.nr_coefficients, self.nr_atom_basis_post_norm_hidden, activation=activation),
+            Dense(self.nr_atom_basis_post_norm_hidden, self.nr_atom_basis_post_norm_out, activation=activation)
+        )
+
+        self.edge_model = ContinuousFilterConvolutionWithConcatenation(self.nr_atom_basis * 2,
+                                                                       self.nr_edge_basis_hidden,
+                                                                       self.nr_edge_basis,
+                                                                       radial_basis_module,
+                                                                       activation)
+
+        self.semantic_attention_mlp = Dense(self.nr_edge_basis, self.nr_heads, activation=nn.CELU(alpha=2.0))
+
+        self.velocity_mlp = nn.Sequential(
+            Dense(self.nr_atom_basis, self.nr_atom_basis_velocity, activation=activation),
+            Dense(self.nr_atom_basis_velocity, 1, activation=lambda x: 2.0 * F.sigmoid(x), bias=False)
+        )
+
+        self.x_mixing_mlp = Dense(self.nr_heads * self.nr_edge_basis, self.nr_coefficients, bias=False, activation=nn.Tanh())
+
+        self.v_mixing = Dense(self.nr_coefficients, 1, bias=False)
+
+    def update_node(self, h, h_i_semantic, h_i_spatial):
+        return h + self.node_mlp(torch.cat([h, h_i_semantic, h_i_spatial], dim=-1))
+
+    def update_velocity(self, v, h, combinations, idx_i):
+        v_ij = self.v_mixing(combinations.transpose(-1, -2)).squeeze(-1)
+        expanded_idx_i = idx_i.view(-1, 1).expand_as(v_ij)
+        dv = torch.zeros_like(v).scatter_reduce(0, expanded_idx_i, v_ij, "mean", include_self=False)
+        return self.velocity_mlp(h) * v + dv
+
+    def get_combinations(self, h_ij_semantic, dir_ij):
+        return torch.einsum("px,pc->pcx", dir_ij, self.x_mixing_mlp(h_ij_semantic))
+
+    def get_spatial_attention(self, combinations, idx_i, dir_ij, nr_atoms):
+        expanded_idx_i = idx_i.view(-1, 1, 1).expand_as(combinations)
+        out_shape = (nr_atoms, self.nr_coefficients, dir_ij.shape[-1])
+        zeros = torch.zeros(out_shape, dtype=combinations.dtype, device=combinations.device)
+        combinations_mean = zeros.scatter_reduce(0, expanded_idx_i, combinations, "mean", include_self=False)
+        combinations_norm_square = (combinations_mean ** 2).sum(dim=-1)
+        return self.post_norm_mlp(combinations_norm_square)
+
+    def aggregate(self, h_ij_semantic, idx_i, nr_atoms):
+        expanded_idx_i = idx_i.view(-1, 1).expand_as(h_ij_semantic)
+        out_shape = (nr_atoms, self.nr_heads * self.nr_edge_basis)
+        zeros = torch.zeros(out_shape, dtype=h_ij_semantic.dtype, device=h_ij_semantic.device)
+        return zeros.scatter_add(0, expanded_idx_i, h_ij_semantic)
+
+    def get_semantic_attention(self, h_ij_edge, idx_i, d_ij, nr_atoms):
```

Comment:
This is pretty dense code. I think comments and docstrings would help the reader to grasp what is going on here better.